add nix flake

getchoo opened the pull request:
this allows for nix users to easily install albafetch, have access to a development shell in order to contribute, and also adds statically linked binaries for x64/aarch64 in github actions (note: android builds could also be implemented or the current aarch64 builds might work for it, but i don't have a device to test it on)

the only real catch to this is that i had a lot of issues when building with the makefile. i'm not exactly sure what's wrong, but it seems like certain flags being passed to it weren't being used correctly. i tried to manually set flags (nix usually does that automatically, and it's worked flawlessly with other makefiles in my experience) but it still wouldn't be able to link some libraries. after getting a bit frustrated, i decided to make a meson script and after everything worked fine. i didn't want to replace your current makefile, so i just made it a patch that's applied during the nix build -- if you're open to replacing the makefile with it though, i could expand it to be at feature parity, as it's already pretty close and supports linux, darwin, and android

alba4k commented:
yes having feature parity between the Makefile and the meson script would be nice, and it would still be an option to keep both

I was considering cmake, but never really did that (am too lazy)

I will check whether the arm64 static binary works on android this evening, although I doubt it will (the libraries that it's statically built with are GNU/Linux ones (the gnu clarification is relevant here, since we are indeed talking about android). I'm especially sceptic on pci.h

the build I was planning to put in the release would have been built using termux

getchoo commented:
> yes having feature parity between the Makefile and the meson script would be nice

i'll work on that. i'm a bit of a meson noob but since the makefile is simple (and meson itself is pretty easy to use), i'll be able to do it lol

> Edit: it doesn't seem to work in termux.

that was a long shot but i kinda had hope :(
i could try android builds still though. i'll push to another branch now and see what happens

alba4k commented:
Maybe add instructions for the meson build aside from the make ones in README.md

getchoo commented:
alright, after a lot of force pushing and doc reading, darwin builds seem good to go. i fixed issues with meson not finding a compiler for objc see (just needed `add_languages('objc')` 🤦🏻), fixed an issue with IOKit not being included (which would lead to kIOMainPortDefault not being declared) and allowed for the conditional use of `kIOMasterPortDefault` when building against older sdks

i'm not sure how android builds are going since i can't test, so i'm not including them in this pr -- builds are available [here](https://github.com/getchoo/albafetch/actions/runs/4548029016) and [here](https://github.com/getchoo/albafetch/actions/runs/4548791198) if you still want to test it though, since it would be kind of cool but not really a big deal if they're missing

i'll add meson instructions to the readme now, and afterwards this should be ready for merge

getchoo commented:
also added instructions for building/installing with nix since those are options now

alba4k commented:
I'm merging, let's see how this goes :)

alba4k commented:
@getchoo Any clue why [this line](https://github.com/alba4k/albafetch/blob/fbbed48e14d053fa1b5f10170ce94f581e710bd0/src/macos/macos_gpu_string.m#L13)
```c
#if __OSX_AVAILABLE_STARTING(__MAC_12_0,__IPHONE_NA)
```
is now causing ci to fail (see [here](https://github.com/alba4k/albafetch/actions/runs/12534789837/job/34955969468))

getchoo commented:
I'm guessing it may be due to the `__MAC_12_0` symbol missing, as Nixpkgs uses v11 of the SDK by default.

This patch should make it use v12 where this will definitely be. I recommend trying to find a way to build this on the current version though (as this is a sign compilation will fail on any macOS version < 12)

```patch
diff --git a/nix/package.nix b/nix/package.nix
index 6cc1e2e..cad9187 100644
--- a/nix/package.nix
+++ b/nix/package.nix
@@ -1,6 +1,7 @@
 {
   lib,
   stdenv,
+  apple-sdk_12,
   meson,
   ninja,
   pciutils,
@@ -35,9 +36,12 @@ stdenv.mkDerivation {

   src = self;

-  buildInputs = [
-    sqlite
-  ] ++ lib.optional stdenv.hostPlatform.isLinux pciutils;
+  buildInputs =
+    [
+      sqlite
+    ]
+    ++ lib.optional stdenv.hostPlatform.isDarwin apple-sdk_12
+    ++ lib.optional stdenv.hostPlatform.isLinux pciutils;

   nativeBuildInputs = [
     meson
```
